Sources/SwiftSyntax/WeakLookupTable.swift

```diff
@@ -0,0 +1,213 @@
+    // or the bucket holding the id.
+    while true {
+      guard let obj = buckets[bucket].value else {
+        return (bucket, false)
+      }
+      if obj.id == id {
+        return (bucket, true)
+      }
+      bucket = (bucket &+ 1) & _bucketMask
+    }
+  }
+
+  /// Reserves enough space to store the specified number of elements. Returns
+  /// true if resizing happened.
+  func reserveCapacity(_ requiredCapacity: Int) -> Bool {
+    let requiredBucketCount = WeakLookupTable<Element>
+      ._bucketCount(for: requiredCapacity, from: bucketCount)
+    if (bucketCount >= requiredBucketCount) {
+      return false
+    }
+
+    // Slow path. Resizing.
+    let oldBuckets = buckets
+    let oldBucketRange = buckets ..< buckets.advanced(by: bucketCount)
+
+    bucketCount = requiredBucketCount
+    buckets = .allocate(capacity: requiredBucketCount)
+    buckets.initialize(repeating: nil, count: requiredBucketCount)
+
+    // Move all nodes from the old buffer.
+    for oldBucket in oldBucketRange {
+      if let id = oldBucket.pointee.value?.id {
+        let newBucket = buckets.advanced(by: _findHole(id).pos)
+        newBucket.moveAssign(from: oldBucket, count: 1)
+      } else {
+        oldBucket.deinitialize(count: 1)
+      }
+    }
+
+    oldBuckets.deallocate()
+
+    return true
+  }
+
+  /// Count the actual number of occupied buckets.
+  @inline(__always)
+  private func _countOccupiedBuckets() -> Int {
+    var count = 0
+    for i in 0 ..< bucketCount where buckets[i].value != nil {
+      count &+= 1
+    }
+    return count
+  }
+
+  /// Reserves enough space to store a single new object. Returns true if
+  /// resizing happened.
+  private func _ensurePlusOneCapacity() -> Bool {
+    if bucketCount >= WeakLookupTable<Element>
+                        ._minimalBucketCount(for: estimatedCount &+ 1) {
+      return false
+    }
+
+    // Slow path.
+    estimatedCount = _countOccupiedBuckets()
+    return reserveCapacity(estimatedCount &+ 1)
+  }
+
+  /// Inserts the given object into the table.
+  @discardableResult
+  func insert(_ obj: Element) -> Bool {
+    var (pos, found) = _findHole(obj.id)
+    if found {
+      return false
+    }
+
+    if /*resized=*/_ensurePlusOneCapacity() {
+      pos = _findHole(obj.id).pos
+    }
+    buckets[pos].value = obj
+    estimatedCount &+= 1
+    return true
+  }
+
+  /// Get a object with specified id. Returns 'nil' if the object hasn't been
+  /// insert()-ed or it's already been freed.
+  public subscript(id: Element.Identifier) -> Element? {
+    // Since we don't fill the bucket when the object is freed (because we don't
+    // know), we can't stop iteration at a hole. So in the worst case (i.e. if
+    // the object doesn't exist in the table), full linear search is needed.
+    // However, since we assume the object exists and hasn't been freed yet,
+    // we expect it's stored near the 'idealBucket' anyway.
```

Comment:
We could think about adding a tombstone property to `WeakReference` so that we can detect when we found a hole. 
I.e.
WeakReference starts of as being `{occupied: false, value: nil}`. Once a value is set it becomes `{occupied: true, value: someValue}`. When `someValue` is freed it becomes `{occupied: true, value: nil}`. That way we can stop the search whenever we reach a `WeakReference` with `occupied == false` and thus avoid the full linear search.

However, in our use case, as you mentioned, we always expect to find the element, I think its worth it to make the trade-off to save an extra bit/byte for a linear worst-case complexity that we should never encounter in practice.

TL;DR: No need to change anything unless we are planning to merge this into `corelibs-foundation`.

Reply:
Yeah, I don't intend to merge this into `corelibs-foundation`. :)
`NSHashTable` has very different API anyway.